Library/Homebrew/cmd/uninstall.rb

```diff
@@ -26,6 +53,36 @@ def uninstall
         name = Formulary.canonical_name(name)
         rack = HOMEBREW_CELLAR/name
 
+        # Cellar/oldname can be a symlink to Cellar/newname so in order not to
+        # get newname's keg as oldname's keg we check if it's a symlink.
+        if rack.symlink?
+          raise "#{rack.to_s} is a symlink"
+        end
```

Comment:
> Why half of files? It doesn't uninstall anything

Sorry, I miss-read the codes. 

Actually, I think you don't need to check the symlink. Noted `name = Formulary.canonical_name(name)`, if you change the `Formulary` logic as we discussed, `canonical_name` should return the proper name(i.e. not old name).